src/Interpreters/InterpreterSelectQuery.cpp

```diff
@@ -1832,7 +1832,10 @@ void InterpreterSelectQuery::executeFetchColumns(QueryProcessingStage::Enum proc
     if (storage && storage->isRemote())
     {
         is_remote = true;
-        max_threads_execute_query = max_streams = settings.max_distributed_connections;
+        if (!settings.async_socket_for_remote)
+            max_threads_execute_query = max_streams = settings.max_distributed_connections;
+        else
+            max_streams = std::min<size_t>(max_streams, settings.max_distributed_connections);
```

Comment:
But what this changes?
I mean that `max_distributed_connections` will not limit number of connections even with this patch:

```sql
SELECT sleep(1)
FROM remote('127.{2,3}', system.one)
SETTINGS max_distributed_connections = 1

Query id: 01a64150-456c-4b0c-9677-ee2e447780aa

┌─sleep(1)─┐
│        0 │
└──────────┘
┌─sleep(1)─┐
│        0 │
└──────────┘

2 rows in set. Elapsed: 1.008 sec. 

SELECT sleep(1)
FROM remote('127.{2,3}', system.one)
SETTINGS max_distributed_connections = 1, max_threads = 1

Query id: e2f633d6-dc1e-4c3c-a2e5-29b7b8cce672

┌─sleep(1)─┐
│        0 │
└──────────┘
┌─sleep(1)─┐
│        0 │
└──────────┘

2 rows in set. Elapsed: 1.009 sec. 
```

And AFAICS `max_streams` is passed to the `IStorage::read`, which will be ignored in `StorageDistributed::read` eventually.